### PR TITLE
Config api

### PR DESCRIPTION
**Pull request by Ian-Stewart-Binks**

### CODE REVIEW

**Comment by Ian-Stewart-Binks** [comment]
@fbricon I've made the config-api branch mergable. It's probably a good idea to review these changes once again, just in case we've developed better ideas since the changes first went in. Do you have any thoughts? Thanks.

It looks like the `org.eclipse.buildship.jdt` plug-in hasn't made its way into Buildship yet. Perhaps we should create one for the JDT configurator (or at least, a plug-in with a different name, perhaps `org.eclipse.buildship.configurator.jdt`, that would introduce a new naming convention that we may not desire, however.)

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java [hunk 1]
```diff
@@ -35,6 +39,7 @@
 
     private final WorkspaceOperations workspaceOperations;
     private final ProjectConfigurationPersistence projectConfigurationPersistence;
+    private IProjectConfigurator[] projectConfigurators = {new JavaProjectConfigurator()};
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: Take this out, need to implement extension point reader.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java [hunk 2]
```diff
@@ -101,4 +106,27 @@ public ProjectConfiguration readProjectConfiguration(IProject workspaceProject)
...
+        MultiStatus multiStatus = new MultiStatus(CorePlugin.PLUGIN_ID, IStatus.OK, "", null);
+
+        for (IProjectConfigurator projectConfigurator : this.projectConfigurators) {
+            if (projectConfigurator.canConfigure(configurationRequest)) {
+                IStatus result = projectConfigurator.configure(configurationRequest, monitor);
+                if (!result.isOK()) {
+                    multiStatus.add(result);
+                }
+            } else {
+                monitor.worked(1);
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: I don't have a good understanding of how monitors work. I'll need to make sure that the code is using monitors correctly.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/IProjectConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+package org.eclipse.buildship.core.configuration.internal;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+
+/**
+ * A Project Configurator.
+ * @author istewart
+ * @since 1.0.1
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/IProjectConfigurator.java]: Buildship doesn't seem to use this convention, would likely be removed.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/IProjectConfigurator.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+     * @param configurationRequest
+     * @return Whether the given project can be configured.
+     */
+    public boolean canConfigure(ProjectConfigurationRequest configurationRequest);
+
+    /**
+     * Configures a project.
+     * @param workspaceOperations
+     * @param modelProject
+     * @param monitor
```
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/IProjectConfigurator.java]: Adds nothing to documentation.

**Comment by fbricon** [comment]
Would be nice if org.eclipse.buildship.jdt was available for sure. For now we can keep the java configurator in core to limit the area of changes in the file structure and focus on the javaee part.

But yeah, the java configurator should be moved out, eventually

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java [hunk 5]
```diff
@@ -101,4 +106,27 @@ public ProjectConfiguration readProjectConfiguration(IProject workspaceProject)
...
+    /**
+     * Updates the configuration of a given project.
+     * @param workspaceProject
+     * @param project
+     * @param monitor
+     */
+    public IStatus updateConfiguration(ProjectConfigurationRequest configurationRequest, IProgressMonitor monitor) {
+        MultiStatus multiStatus = new MultiStatus(CorePlugin.PLUGIN_ID, IStatus.OK, "", null);
+
+        for (IProjectConfigurator projectConfigurator : this.projectConfigurators) {
```
Comment (fbricon) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: generally check if monitor is cancelled, for each loop

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java (hunk 2 above)
Comment (fbricon) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: http://www.vogella.com/tutorials/EclipseJobs/article.html#eclipsejobs_iprogressmonitor

Basically you can prepare the monitor.beginTask("Configure project", numConfigurators); and do monitor.worked(1) in each loop (whether the configurator was invoked or not);

Comment (fbricon) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: move monitor.worked out of else clause

Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: Done. Thanks for the link.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java (hunk 5 above)
Comment (Ian-Stewart-Binks) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/configuration/internal/DefaultProjectConfigurationManager.java]: Done. I'm creating a new `MultiStatus` with the `IStatus.CANCEL` code.

**Pull request merged**